### DISCUSSION TITLE
General design discussion

### DISCUSSION THREAD

**Discussion by RetricSu** [ckb-devrel/offckb#3] [Ideas]
see https://github.com/RetricSu/offckb/discussions/3#discussioncomment-8631900 for updates
---
## Purpose

The general idea is to use `offckb` to quickly provide an easy clean and full local CKB development environment for dapp developers(mainly for the js/ts tech stack). and starting from here, we may add more tooling and features for contract developers. But the js/web developers are the main ones at the first stage.

To do that, we need to cover serval things:

1. A tool/shell script to easily run CKB devnet on a local environment, and this chain should include some most useful smart contracts (scripts).
2. A fixed account list, pre-funded with devnet tokens.
3. A way to view and check the local blockchain status
4. Integrated with DAPP framework like [lumos](https://github.com/ckb-js/lumos) to bring a unified blazing fast development experience.

## MVP

To start with, we will build:

1. A CLI tool can be installed via `npm install` and executed to 
  - start the devnet
  - check account balance
  - query chain-info
  - and most importantly, to integrate with `lumos` to init and build the CKB DAPP project in typescript. Think `hardhat` for Ethereum.
  
```js
npm install -g offckb
offckb node // start the devnet of CKB 
offckb init // init a typescript boilerplate with lumos to get started with to build CKB DAPP like hardhat init
offckb list-accounts
offckb list-hashes // list scripts hashes, equals `ckb list-hashes`

//.... maybe even more features like
offckb deploy
offckb test
```

2. a shell script in one-liner style(hopefully) to use pre-build docker image or docker-compose to start the devnet.

```sh
docker run offckb:v0.1.0

# or use docker-compose
git clone https://github.com/RetricSu/offckb.git && cd offckb/docker && docker-compose up -d
```

**Comment by RetricSu** [comment]
### Task

**First stage**

- [x] build a docker-compose with devnet
- [x] find a way to embed the scripts into devnet config
- [x] build a cli `offckb` to run the devnet
- [x] integrated `offckb` with lumos to leverage this devnet
  - [x] auto gen lumos config from this devnet
  - [x] init a typescript boilerplate with lumos
- [x] config account-list with pre-fund tokens
- [x] setup project CI
  - [x] linting/building 
  - [x] publish the cli to NPM with CI
- [x] testing
  - [x] test the whole process
  - [x] test built-in scripts usage to see if it works expected （thx for acceptance team)
  - [x] gain feedbacks
- [x] ~build a docker-image with devnet~

**Second stage**

- [x] support windows
- [x] support clean node
- [x] improve simple UX on the transfer template
  from @yfeng2824 :
  - [x] Maybe we should disable the "Transfer" button before users fill out all 3 inputs.
  - [x] Instead of saying "too short" as the error message, displaying a more actionable direction will be better. For example, we can display "Invalid private key: Ensure you're using a valid private key from the offckb accounts list."
  - [x] I was trying to transfer all CKB from one address to another, and a modal pops out saying "Error: Not enough CKB". Maybe we should tell users that transferring will cost extra, so that users can adjust the amount accordingly.  Ideally, we should display insufficient CKB as the error message before users hit Transfer.
  - [x] I think format the large number with commas can provide better readability. For example, we can format capacity "4200000000000000" to "4,200,000,000,000,000"
- [x] better docs with built-in scripts version and Spore/Lumos docs link
- [x] add `spore` as built-in script
- [x] write `xudt` templates
- [x] write `spore` NFT templates
- [ ] write `omnilock` templates
- [ ] write `anyone-can-pay` templates

**Comment by Flouse** [comment]
> serval things:
> 
> 1. A tool/shell script to easily run CKB devnet on a local environment, and this chain should include some most useful smart contracts (scripts).
> 2. A fixed account list, pre-funded with devnet tokens.
> 3. A way to view and check the local blockchain status
> 4. Integrated with DAPP framework like [lumos](https://github.com/ckb-js/lumos) to bring a unified blazing fast development experience.

https://github.com/ckb-js/kuai may support some of the features listed here.
cc @Keith-CY

**Comment by RetricSu** [comment]
> > serval things:
> > 
> > 1. A tool/shell script to easily run CKB devnet on a local environment, and this chain should include some most useful smart contracts (scripts).
> > 2. A fixed account list, pre-funded with devnet tokens.
> > 3. A way to view and check the local blockchain status
> > 4. Integrated with DAPP framework like [lumos](https://github.com/ckb-js/lumos) to bring a unified blazing fast development experience.
> 
> https://github.com/ckb-js/kuai may support some of the features listed here. cc @Keith-CY

cool, haven't heard news from kuai yet for some time. I am curious what is the status of kuai? here we are refactoring the docs site for CKB, and the problem we faced is that asking developers (especially dapp developers focus on js/ts tech stack) to run a devnet involves too many steps(you need to edit a lot of configs), we want to lower the cost here, so we plan to do a very thin wrap around devnet to get developers started. If kuai already cover such thing we are happy to use that. also, another very important question: do we recommend kuai instead of lumos for developers nowadays when they're looking to build on CKB? What are your thoughts about it? @Flouse @Keith-CY

**Reply by Flouse** [reply]
I think Kuai is using lumos too, see also https://github.com/ckb-js/kuai/blob/dbde4593ae8c243210dbb04b7e1e3cd70ab34da4/packages/core/package.json#L16

**Reply by Keith-CY** [reply]
In the Q4 of 2023, the team of kuai has built an demo of sudt management at https://sudt-management-ui-production.up.railway.app/ which can be connected to Neuron via Wallet Connect(https://github.com/nervosnetwork/neuron/actions/runs/7057329140). But the solution with Wallet Connect was discarded after co-build protocol was proposed, so the demo was not promoted.

Kuai keeps going when we have proper developers who are good at framework development. The model of kuai framework is functional, composable but its mental model is a bit different from that used by App developers.

---

BTW
- Running a local devnet with built-in scripts(https://github.com/ckb-js/ckb-production-scripts/releases) and user-defined scripts: https://github.com/ckb-js/kuai/blob/develop/packages/core/src/builtin-tasks/node.ts#L94 has been supported
- A default account can be set in the configuration(https://github.com/ckb-js/kuai/blob/develop/packages/core/src/type/runtime.ts#L10) and a list of accounts can be set on running a node to distribute dev tokens(https://github.com/ckb-js/kuai/blob/develop/packages/core/src/builtin-tasks/node.ts#L38)
- CKB Explorer is going to support connecting to full node directly(https://github.com/Magickbase/ckb-explorer-public-issues/issues/526), by doing so the status of local node can be visualized.

**Comment by RetricSu** [comment]
We have a MVP https://www.npmjs.com/package/@offckb/cli 

I only tested it on Mac, Linux should work too, Windows is not supported now. Also, the docker version of this is not supported either.

looking for some feedback.

**Comment by RetricSu** [comment]
Given the feedback, we will adjust the `offckb` position with a more clear vision:

Instead of making `offckb` the `hardhat of CKB`, we will focus more on building `offckb` to better assist with the new Nervos docs site. That means instead of providing a powerful and flexible Dapp development environment, we will minimize the entrance cost for first-time developers and hardcode more Dapp templates to serve as educational tutorials.

That brings two adjustments:

1. Remove the docker node support to run the local devnet. This helped lower the cost for users and it is also easy for us to maintain this project.
2. `offckb init` with multiple Dapp templates to showcase the most important and latest scripts in CKB.

--- 

In my opinion, there are two big problems for Dapp developers entering the CKB world:

The first one is onboarding, which results from poor-docs/hard-to-understanding of the design pattern of the UTXO model. The second one is lacking of well-designed easy-to-setup Dapp development tooling and infrastructure compared to ecosystems like Ethereum.

`kuai` has the ambition to solve the second problem. It covers the full life cycle of a Dapp from starting the local development environment to writing tests and deploying the scripts. Besides that, it also aims to introduce a new UTXO Dapp development pattern inspired by the Actor model. We are solving the first problem while leaving the second problem to `kuai`, hopefully.
